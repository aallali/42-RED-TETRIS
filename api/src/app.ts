
import * as express from "express";
import * as cors from "cors"
import { Socket } from "socket.io";
import Game from "./Game"

import ROOM from "./Room"
import PLAYER from "./Player"
import { ROOM_MODE } from "./types"


const log = console.log
interface IPlayer {
	id: string;
	nickname: string;
}
interface IRoom {
	id: string;
	name: string;
	players: []
	admin: IPlayer
}
class App {

	public app: express.Application;
	public http: any
	public io: Socket
	private port: number = 4242
	private gameIns: Game
	constructor() {
		this.app = express();
		this.app.use(
			cors({
				origin: "*",
				credentials: true,
			})
		);
		this.http = require("http").Server(this.app);
		this.io = require("socket.io")(this.http, {
			pingInterval: 60000,
			cors: {
				origin: "*",
			},
		});
		this.routes()
		this.initSocketEvents()
		this.gameIns = new Game(this.io)

	}
	/**
	 * 
	 */
	private routes() {
		this.app.get("/", (req: any, res: any) => {
			console.log(Array.from(this.gameIns.players))
			res.send(Array.from(this.gameIns.players))
		});
	}
	private initSocketEvents() {
	}
	public listen() {
		console.clear()
		this.http.listen(this.port, () =>
			console.log(`ðŸš€ Server is running on PORT ${this.port}`)
		);
	}
}

export default new App();
