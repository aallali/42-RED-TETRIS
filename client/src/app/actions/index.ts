import {
	SET_PLAYER, SET_GAME,
	LOGOUT_PLAYER, UPDATE_SCORE,
	SET_PLAYER_ADMIN, ADD_ROW,
	PLAYER_LOST, RESET_STATES,
	BROADCAST_SCORE,
	UNSET_PLAYER_ADMIN
} from "../../reducers/player.reducer";
import { UPDATE_PLAYERS, CLEAR_OPPONENTS } from "../../reducers/opponent.reducer"
import { SET_ERROR, CLEAR_ERROR } from "../../reducers/error.reducer"

import {
	SHIFT_TETRO, MORE_TETROS,
	SET_GAME_MODE, SET_GAME_OVER,
	SET_GAME_STARTED, SET_IN_GAME,
	SET_WINNER, RESET_GAME,
	UPDATE_GAME_MODE,
	LEAVE_GAME,
	SET_GAME_TITLE
} from "../../reducers/game.reducer"

export {
	// Players actions
	SET_PLAYER, SET_GAME,
	LOGOUT_PLAYER,
	UPDATE_SCORE,
	SET_PLAYER_ADMIN,
	ADD_ROW, PLAYER_LOST,
	BROADCAST_SCORE,
	UNSET_PLAYER_ADMIN,
	RESET_STATES,

	// Opponents actions
	UPDATE_PLAYERS,
	CLEAR_OPPONENTS,
	// Error actions
	SET_ERROR, CLEAR_ERROR,

	// game actions
	SHIFT_TETRO, MORE_TETROS,
	SET_GAME_MODE, SET_GAME_OVER,
	SET_GAME_STARTED, SET_IN_GAME,
	SET_WINNER, RESET_GAME,
	UPDATE_GAME_MODE,
	LEAVE_GAME,
	SET_GAME_TITLE
}
