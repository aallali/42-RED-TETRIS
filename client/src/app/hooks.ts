import { TypedUseSelectorHook, useDispatch, useSelector } from 'react-redux';
import type { RootState, AppDispatch } from './store';
import { useGameStatus } from "./TetrisHooks/useGameStatus"
import { useInterval } from "./TetrisHooks/useInterval"
import { usePlayer } from "./TetrisHooks/usePlayer"
import { useStage } from "./TetrisHooks/useStage"


import io from "socket.io-client";

export const socket = io("http://localhost:4242");
export { useGameStatus, usePlayer, useInterval, useStage }
// Use throughout your app instead of plain `useDispatch` and `useSelector`
export const useAppDispatch = () => useDispatch<AppDispatch>();
export const useAppSelector: TypedUseSelectorHook<RootState> = useSelector;
