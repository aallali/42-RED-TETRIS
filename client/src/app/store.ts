import { configureStore, ThunkAction, Action } from '@reduxjs/toolkit';
import playerReducer from "../reducers/player.reducer";
import stageReducer from "../reducers/stage.reducer"
export const store = configureStore({
  reducer: {
    player: playerReducer,
    stage: stageReducer
  },
});

export type AppDispatch = typeof store.dispatch;
export type RootState = ReturnType<typeof store.getState>;
export type AppThunk<ReturnType = void> = ThunkAction<
  ReturnType,
  RootState,
  unknown,
  Action<string>
>;
