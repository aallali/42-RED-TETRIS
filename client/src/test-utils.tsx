// test-utils.jsx
import React, { ReactChildren } from 'react'
import { render as rtlRender } from '@testing-library/react'
import { configureStore } from '@reduxjs/toolkit'
import { Provider } from 'react-redux'
// Import your own reducer
import playerReducer from "./reducers/player.reducer";
import opponentReducer from "./reducers/opponent.reducer"
import errorReducer from "./reducers/error.reducer"
import gameReducer from "./reducers/game.reducer"
import chatReducer from './reducers/chat.reducer';

function aallaliStore(preloadedState: any) {
	return configureStore({
		reducer: {
			player: playerReducer,
			opponents: opponentReducer,
			error: errorReducer,
			game: gameReducer,
			chat: chatReducer
		},
		preloadedState
	})
}
function render(
	ui: React.ReactElement<any, string | React.JSXElementConstructor<any>>,
	{
		preloadedState,
		store = configureStore({
			reducer: {
				player: playerReducer,
				opponents: opponentReducer,
				error: errorReducer,
				game: gameReducer,
				chat: chatReducer
			},
			preloadedState
		}),
		...renderOptions
	}: any = {}
) {
	function Wrapper({ children }: { children: ReactChildren }) {
		return <Provider store={store}>{children}</Provider>

	}
	return rtlRender(ui, { wrapper: Wrapper, ...renderOptions })
}


// re-export everything
export * from '@testing-library/react'
// override render method
export { render, aallaliStore }