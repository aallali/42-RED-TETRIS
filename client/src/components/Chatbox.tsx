
/* eslint-disable react-hooks/exhaustive-deps */

import { createRef, useEffect, useState } from "react";
import { socket, useAppDispatch, useAppSelector } from "../app/hooks";
import { ADD_MSG, getChatMsgs } from "../reducers/chat.reducer";
import { getOpponents } from "../reducers/opponent.reducer";
import "./css/chatbox.css"

const SentMsg = ({ msg }: { msg: string }) => (<div className="message bg-black text-white p-1 self-end my-1 rounded-md shadow ml-3">
	{msg}
</div>)
const ReveicedMsg = ({ from, msg }: { from: string, msg: string }) => (<div className="message bg-white text-gray-700 p-1 self-start my-1 rounded-md shadow mr-3">
	<b>[{from}]</b>:{msg}
</div>)
function ChatBox() {
	const dispatch = useAppDispatch()
	const boxRef = createRef<any>()
	const friends = useAppSelector(getOpponents).map(l => l.name)
	const [msg, setMsg] = useState<string>("")
	const chatMsgs = useAppSelector(getChatMsgs)
	const [newMsgs, setNewMsg] = useState(false)
	const [chatOpen, setChatOpen] = useState(false)
	const handleMsg = (v: string) => {
		if (v?.length <= 100)
			setMsg(v)
	}

	const handleSendMsg = () => {
		if (msg?.trim().length > 0) {
			dispatch(ADD_MSG({ from: "sssss", msg: msg }))
			socket.emit("CHAT_MSG", { msg: msg })
			setMsg("")
		}
	}

	const handleKeyDown = ({ key }: { key: string }) => {
		if (key === 'Enter') {
			handleSendMsg()
		}
	}

	const closeModel = () => {
		const chatModal: Element | null = document.querySelector('.chat-modal');
		const chatServices: Element | null = document.querySelector('.chat-services');
		const showChat: Element | null = document.querySelector('.show-chat');

		setTimeout(() => {
			showChat?.classList.remove('hidden')
		}, 820);
		chatServices?.classList.remove('expand')
		setTimeout(() => {
			chatModal?.classList.remove('show')
		}, 500);
		setChatOpen(false)
		setNewMsg(false)
		return true
	}

	const showModel = () => {
		const chatModal: Element | null = document.querySelector('.chat-modal');
		const chatServices: Element | null = document.querySelector('.chat-services');
		const showChat: Element | null = document.querySelector('.show-chat');

		chatModal?.classList.add('show')
		showChat?.classList.add('hidden')
		setTimeout(() => {
			chatServices?.classList.add('expand')
		}, 500);
		setNewMsg(false)
		setChatOpen(true)
		return true
	}
	useEffect(() => {
		if (boxRef.current)
			boxRef.current.scrollTop = boxRef.current.scrollHeight;
		setNewMsg(true)
	}, [chatMsgs])

	useEffect(() => {
		// closeModel()
		setTimeout(() => {
			showModel()
		}, 500)
	}, [])
	useEffect(() => {
		const showChat: Element | null = document.querySelector('.show-chat');
		if (newMsgs && !chatOpen) {
			showChat?.classList.add('blob')
			showChat?.classList.add('red')
		}
		else {
			showChat?.classList.remove('blob')
			showChat?.classList.remove('red')
		}
	}, [newMsgs, chatOpen])
	return (
		<div>

			<div className="fixed bottom-0 right-0 flex flex-col items-end ml-6  w-auto ">
				<div className={"chat-modal mr-1 flex flex-col mb-5 shadow-lg w-full"}
					data-testid="chat-modal">
					<div
						className="close-chat bg-red-500 hover:bg-red-600 text-white mb-1 w-10 flex justify-center items-center px-2 py-1 rounded self-end cursor-pointer"
						data-testid="closeModal"
						onClick={() => closeModel()}
					>
						<svg width="1em" height="1em" viewBox="0 0 16 16" className="bi bi-x" fill="currentColor"
							xmlns="http://www.w3.org/2000/svg">
							<path fillRule="evenodd"
								d="M11.854 4.146a.5.5 0 0 1 0 .708l-7 7a.5.5 0 0 1-.708-.708l7-7a.5.5 0 0 1 .708 0z" />
							<path fillRule="evenodd"
								d="M4.146 4.146a.5.5 0 0 0 0 .708l7 7a.5.5 0 0 0 .708-.708l-7-7a.5.5 0 0 0-.708 0z" />
						</svg>
					</div>

					<div
						className="flex justify-between items-center text-white p-1 pl-4 bg-black border shadow-lg mr-5 w-full rounded-t-lg">
						<div className="flex items-center">
							{
								friends.map(l => (<img
									src={"https://ui-avatars.com/api/?background=B82E1F&bold=true&color=fff&name=" + l}
									alt={l}
									key={l}
									className="rounded-full w-8 h-8 mr-1 -ml-3 border-2 border-white">
								</img>))
							}

							<h2 className="font-semibold tracking-wider">{friends.join(',')}</h2>

						</div>

					</div>

					<div className={"flex flex-col bg-gray-200 px-2 chat-services overflow-auto "} ref={boxRef}
						data-testid="chat-services"
					>
						{
							chatMsgs.map((l, i) => {
								if (!l.from) return <SentMsg key={i} msg={l.msg} />
								return <ReveicedMsg key={i} from={l.from} msg={l.msg} />
							})
						}
					</div>
					<div className="relative bg-white ">
						<input type="text" name="message" placeholder="write your message here ..."
							className="pl-4 pr-16 py-2 border border-black-500 focus:outline-none w-full"
							value={msg}
							onChange={(e) => handleMsg(e.target.value)}
							onKeyDown={handleKeyDown}
						>
						</input>
						<button
							className="absolute right-0 bottom-0 text-black bg-white  hover:text-blue-500 m-1 px-3 py-1 w-auto transistion-color duration-100 focus:outline-none"
							data-testid="send"
							onClick={() => handleSendMsg()}
						>Send
						</button>
					</div>
				</div>
				<div
					className={"show-chat mx-10 mb-6 mt-4 text-white hover:text-gray-500 flex justify-center items-center cursor-pointer "}
					data-testid="show-chat"
					onClick={() => showModel()}
				>
					<svg width="4em" height="2em" viewBox="0 0 16 16" className="bi bi-chat-text-fill" fill="currentColor"
						xmlns="http://www.w3.org/2000/svg">
						<path fillRule="evenodd"
							d="M16 8c0 3.866-3.582 7-8 7a9.06 9.06 0 0 1-2.347-.306c-.584.296-1.925.864-4.181 1.234-.2.032-.352-.176-.273-.362.354-.836.674-1.95.77-2.966C.744 11.37 0 9.76 0 8c0-3.866 3.582-7 8-7s8 3.134 8 7zM4.5 5a.5.5 0 0 0 0 1h7a.5.5 0 0 0 0-1h-7zm0 2.5a.5.5 0 0 0 0 1h7a.5.5 0 0 0 0-1h-7zm0 2.5a.5.5 0 0 0 0 1h4a.5.5 0 0 0 0-1h-4z" />
					</svg>
				</div>
			</div>

		</div>
	)
};

export default ChatBox
