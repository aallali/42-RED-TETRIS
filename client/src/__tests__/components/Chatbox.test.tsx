import renderer, { act } from "react-test-renderer";
import Chatbox from "../../components/Chatbox"
import { Provider } from 'react-redux';
import * as reactRedux from "react-redux";
import { render, fireEvent, screen } from '../../test-utils'
import { store } from "../../app/store"
// so we can import fireEvent and screen here as well
jest.setTimeout(10 * 1000)
describe("_ ChatBox Component test", () => {
	const useDispatchMock = jest.spyOn(reactRedux, "useDispatch");
	beforeEach(() => {
		useDispatchMock.mockClear();
	});


	test("test chatbox rendering ", () => {
		const tree = renderer.create(<Provider store={store}>
			<Chatbox />
		</Provider>).toJSON();
		expect(tree).toMatchSnapshot();
	});

	it("simulate ENTER button click in chatbox input ", () => {
		// const renderer = new ShallowRenderer();
		// const wrapper = mount(<Provider store={store}>
		// 	<Chatbox />
		// </Provider>)
		// // wrapper.render()
		// const instance = wrapper.instance();
		// console.log(instance.handleKeyDown({key:"Enter"}))
		// wrapper.find('[placeholder="write your message here ..."]').simulate('keypress', {key: 'Enter'})
		// wrapper.update()
		// expect(useDispatchMock).toHaveBeenCalled();
		// wrapper.find('[placeholder="write your message here ..."]').simulate('keypress', {key: 'Space'})
		// wrapper.update()

		// expect(wrapper.state('chatopen')).toEqual(true);
		// expect(wrapper.find('.fixed')).toHaveLength(1)
	});

	test('test send message chatbox', async () => {
		useDispatchMock.mockReturnValue(jest.fn());

		render(
			<Chatbox />, {
			preloadedState:
			{
				opponents: {
					players: [{
						name: "testFriend",
						lost: false,
						stage: [],
						rows: 0,
						score: 0,
						level: 0,
						admin: false
					}]
				},
				chat: { msgs: [{ from: "", msg: "this message is from me" }] }
			}
		}
		)
		jest.useFakeTimers()
		const roomInput = screen.getByPlaceholderText(/write your message here \.\.\./i);

		expect(roomInput).toBeInTheDocument()

		fireEvent.change(roomInput, { target: { value: "Hi, this is my first msg in chatbox" } });
		expect(screen.getByText(/Send/)).toBeInTheDocument()
		expect(screen.getByText(/testFriend/)).toBeInTheDocument()

		fireEvent.click(screen.getByText(/Send/));
		expect(useDispatchMock).toHaveBeenCalled();

		act(() => {
			jest.runOnlyPendingTimers();
		});

		act(() => {
			jest.runAllTimers();
		});
		jest.advanceTimersByTime(1000);

		expect(screen.getByTestId("chat-modal")).toHaveClass('show')
		expect(screen.getByTestId("show-chat")).toHaveClass('hidden')
		expect(screen.getByTestId("chat-services")).toHaveClass('expand')

		act(() => {
			fireEvent.click(screen.getByTestId("closeModal"))
		});

		jest.advanceTimersByTime(1000);
		expect(screen.getByTestId("chat-modal")).not.toHaveClass('show')
		expect(screen.getByTestId("show-chat")).not.toHaveClass('hidden')
		expect(screen.getByTestId("chat-services")).not.toHaveClass('expand')

		act(() => {
			fireEvent.click(screen.getByTestId("show-chat"))
		});

		jest.advanceTimersByTime(1000);
		expect(screen.getByTestId("chat-modal")).toHaveClass('show')
		expect(screen.getByTestId("show-chat")).toHaveClass('hidden')
		expect(screen.getByTestId("chat-services")).toHaveClass('expand')
	})
})

