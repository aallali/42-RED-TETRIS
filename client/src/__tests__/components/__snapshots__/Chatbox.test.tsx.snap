// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`_ ChatBox Component test test chatbox rendering  1`] = `
<div>
  <div
    className="fixed bottom-0 right-0 flex flex-col items-end ml-6  w-auto "
  >
    <div
      className="chat-modal mr-1 flex flex-col mb-5 shadow-lg w-full"
      data-testid="chat-modal"
    >
      <div
        className="close-chat bg-red-500 hover:bg-red-600 text-white mb-1 w-10 flex justify-center items-center px-2 py-1 rounded self-end cursor-pointer"
        data-testid="closeModal"
        onClick={[Function]}
      >
        <svg
          className="bi bi-x"
          fill="currentColor"
          height="1em"
          viewBox="0 0 16 16"
          width="1em"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M11.854 4.146a.5.5 0 0 1 0 .708l-7 7a.5.5 0 0 1-.708-.708l7-7a.5.5 0 0 1 .708 0z"
            fillRule="evenodd"
          />
          <path
            d="M4.146 4.146a.5.5 0 0 0 0 .708l7 7a.5.5 0 0 0 .708-.708l-7-7a.5.5 0 0 0-.708 0z"
            fillRule="evenodd"
          />
        </svg>
      </div>
      <div
        className="flex justify-between items-center text-white p-1 pl-4 bg-black border shadow-lg mr-5 w-full rounded-t-lg"
      >
        <div
          className="flex items-center"
        >
          <h2
            className="font-semibold tracking-wider"
          >
            
          </h2>
        </div>
      </div>
      <div
        className="flex flex-col bg-gray-200 px-2 chat-services overflow-auto "
        data-testid="chat-services"
      >
        <div
          className="message bg-white text-gray-700 p-1 self-start my-1 rounded-md shadow mr-3"
        >
          <b>
            [
            Team (automated)
            ]
          </b>
          :
          Hi 👋! welcome to RED-TETRIS game, 1️⃣ first of all i wish you enjoy the game, 2️⃣ second and lets make it short, you can only write a message under 100 character,when u receive a message, the chatbox icon start an alert animation 🚩 until you open the chatbox, if you notice something to fix feel free to notify me, 3️⃣ third... play 🎮 && have fun  😄
        </div>
        <div
          className="message bg-white text-gray-700 p-1 self-start my-1 rounded-md shadow mr-3"
        >
          <b>
            [
            Team (automated)
            ]
          </b>
          :
          Regards ! Abdellah Allali (hi@allali.me)
        </div>
      </div>
      <div
        className="relative bg-white "
      >
        <input
          className="pl-4 pr-16 py-2 border border-black-500 focus:outline-none w-full"
          name="message"
          onChange={[Function]}
          onKeyDown={[Function]}
          placeholder="write your message here ..."
          type="text"
          value=""
        />
        <button
          className="absolute right-0 bottom-0 text-black bg-white  hover:text-blue-500 m-1 px-3 py-1 w-auto transistion-color duration-100 focus:outline-none"
          data-testid="send"
          onClick={[Function]}
        >
          Send
        </button>
      </div>
    </div>
    <div
      className="show-chat mx-10 mb-6 mt-4 text-white hover:text-gray-500 flex justify-center items-center cursor-pointer "
      data-testid="show-chat"
      onClick={[Function]}
    >
      <svg
        className="bi bi-chat-text-fill"
        fill="currentColor"
        height="2em"
        viewBox="0 0 16 16"
        width="4em"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M16 8c0 3.866-3.582 7-8 7a9.06 9.06 0 0 1-2.347-.306c-.584.296-1.925.864-4.181 1.234-.2.032-.352-.176-.273-.362.354-.836.674-1.95.77-2.966C.744 11.37 0 9.76 0 8c0-3.866 3.582-7 8-7s8 3.134 8 7zM4.5 5a.5.5 0 0 0 0 1h7a.5.5 0 0 0 0-1h-7zm0 2.5a.5.5 0 0 0 0 1h7a.5.5 0 0 0 0-1h-7zm0 2.5a.5.5 0 0 0 0 1h4a.5.5 0 0 0 0-1h-4z"
          fillRule="evenodd"
        />
      </svg>
    </div>
  </div>
</div>
`;
